### ISSUE TITLE
FEValues and FEFaceValues should have symmetric Hessians

### ISSUE THREAD

**Issue by davydden** [dealii/dealii#1554]
subj.

I guess the code was written before `SymmetricTensor` was available.

**Comment by davydden** [comment]
the difference is minor, but casting to SymmetricTensor may fail the internal check, as in this example:
```
(const dealii::Tensor<2, 2, double>) $0 = {
  subtensor = {
    [0] = {
      values = ([0] = 0.51030662532906768, [1] = -0.69755127702650122)
    }
    [1] = {
      values = ([0] = -0.6975512770265011, [1] = -0.6513277984600675)
    }
  }
```

**Comment by bangerth** [comment]
Yes, correct. I thought the same this morning when seeing your change to the cosine function. It's true that the code predates the existence of the SymmetricTensor class. We should eventually fix this.

Of course one could then argue that we need a TotallySymmetricTensor class also for third and higher derivatives so that `t[i][j][k] = t[m][n][p]` where `(mnp)` is any permutation (not necessarily of positive order) or `(ijk)`. But that's beyond my will to add right now :-)

**Comment by davydden** [comment]
:-)

agree, ideally there should be a tensor class of arbitrary order symmetric to arbitrary permutations.

**Comment by bangerth** [comment]
If anyone wants to look at this: The starting point is to change the type of `HessianVector` in `FiniteElementRelatedData` in `fe_update_flags.h`. This implies changes both in the finite element classes as well as the `FEValues` classes, and probably also in the mappings that transform the shape hessians.

There are a number of incompatible downstream changes, for example in `FEValues::get_function_hessians()` that will get a different argument type. It doesn't look like something that's particularly easy to implement.

**Comment by masterleinad** [comment]
Most likely this will not be addressed for 9.0.
